src/backend/replication/walproposer.c

```diff
@@ -150,6 +150,24 @@ CalculateDiskConsistentLsn(void)
 	return lsn;
 }
 
+/*
+ * Get minimum of flushed LSNs of all safekeepers, which is the LSN of the
+ * last WAL record that can be safely discarded.
+ */
+static XLogRecPtr
+CalculateMinFlushLsn(void)
+{
+	XLogRecPtr lsn = UnknownXLogRecPtr;
+	for (int i = 0; i < n_walkeepers; i++)
+	{
+		if (walkeeper[i].feedback.flushLsn < lsn)
```

Comment:
As long as term_history patch is not merged, we also must check here epoch like GetAcknowledgedByQuorumWALPosition does, otherwise flushLsn is essentially lie, it points to wrong history. And if you check the epoch, I believe acknowledgedLsn becomes unnecessary.

(but check `msgQueueHead->ackMask == ((1 << n_walkeepers) - 1)` in queue cleanup is still needed due to 0-sized messages, though it would be nice to simplify this as well)